File: mailblastr-cli/src/commands/templates.js
'use strict';

const { parseJson, clean, withPagination, pagination } = require('../helpers');

/**
 * `variables` is the template's declared-variable registry, not the per-send
 * values: a JSON array of `{"key","type"?,"fallback_value"?}` (max 50, `type`
 * is 'string' | 'number', a number's fallback must parse as a number). The send
 * path prefers a declared fallback when a caller omits the variable, so a
 * template with no registry has no fallbacks — which is why this could not be
 * left to the dashboard. Passing `[]` clears the registry.
 */
const VARIABLES_HELP =
  'declared variables as a JSON array of {"key","type"?,"fallback_value"?} (max 50; [] clears them)';

function register({ group, leaf, act }) {
  const templates = group('templates', 'Manage email templates');

  act(
    leaf(templates, 'create', 'Create a template')
      .requiredOption('--name <name>', 'template name')
      .option('--subject <subject>', 'subject line')
      .option('--html <html>', 'HTML body')
      .option('--text <text>', 'plain-text body')
      .option('--from <from>', 'default sender')
      .option('--reply-to <address>', 'default reply-to')
      .option('--alias <alias>', 'stable handle for sending by alias')
      .option('--variables <json>', VARIABLES_HELP),
    ({ client, opts }) =>
      client.templates.create(
        clean({
          name: opts.name,
          subject: opts.subject,
          html: opts.html,
          text: opts.text,
          from: opts.from,
          reply_to: opts.replyTo,
          alias: opts.alias,
          variables: parseJson(opts.variables, '--variables'),
        }),
      ),
  );

  act(leaf(templates, 'get <id>', 'Retrieve a template'), ({ client, args: [id] }) =>
    client.templates.get(id),
  );

  act(withPagination(leaf(templates, 'list', 'List templates')), ({ client, opts }) =>
    client.templates.list(pagination(opts)),
  );

  act(
    leaf(templates, 'update <id>', "Update a template's draft")
      .option('--name <name>', 'template name')
      .option('--subject <subject>', 'subject line')
      .option('--html <html>', 'HTML body')
      .option('--text <text>', 'plain-text body')
      .option('--from <from>', 'default sender')
      .option('--reply-to <address>', 'default reply-to')
      .option('--alias <alias>', 'stable handle')
      .option('--variables <json>', VARIABLES_HELP),
    ({ client, opts, args: [id] }) =>
      client.templates.update(
        id,
        clean({
          name: opts.name,
          subject: opts.subject,
          html: opts.html,
          text: opts.text,
          from: opts.from,
          reply_to: opts.replyTo,
          alias: opts.alias,
          variables: parseJson(opts.variables, '--variables'),
        }),
      ),
  );

  act(
    leaf(templates, 'duplicate <id>', 'Duplicate a template')
      .option('--name <name>', 'name for the copy')
      .option('--alias <alias>', 'alias for the copy'),
    ({ client, opts, args: [id] }) =>
      client.templates.duplicate(id, clean({ name: opts.name, alias: opts.alias })),
  );

  act(
    leaf(templates, 'publish <id>', 'Publish a template (make its latest draft live)'),
    ({ client, args: [id] }) => client.templates.publish(id),
  );

  act(leaf(templates, 'delete <id>', 'Delete a template'), ({ client, args: [id] }) =>
    client.templates.remove(id),
  );
}

module.exports = { register };

File: mailblastr-python/mailblastr/templates.py
"""Templates resource (incl. duplicate + publish)."""

from typing import Any, Dict, List, TypedDict, Union

from . import http_client
from .http_client import paginate, path_escape as _e

CreateParams = TypedDict(
    "CreateParams",
    {
        "name": str,  # required
        "alias": str,  # optional stable handle for sending by alias
        "subject": Union[str, None],
        "from": Union[str, None],
        "reply_to": Union[str, List[str]],
        "html": Union[str, None],
        "text": Union[str, None],
        "variables": List[Dict[str, Any]],  # [{key, type?, fallback_value?}]
    },
    total=False,
)

UpdateParams = TypedDict(
    "UpdateParams",
    {
        "name": str,
        "alias": str,
        "subject": Union[str, None],
        "from": Union[str, None],
        "reply_to": Union[str, List[str]],
        "html": Union[str, None],
        "text": Union[str, None],
        "variables": List[Dict[str, Any]],
    },
    total=False,
)


class DuplicateParams(TypedDict, total=False):
    name: str
    alias: str


class Templates:
    """``mailblastr.Templates`` — the /templates endpoints."""

    CreateParams = CreateParams
    UpdateParams = UpdateParams
    DuplicateParams = DuplicateParams

    @classmethod
    def create(cls, params):
        """Create a template. POST /templates"""
        return http_client.request("POST", "/templates", params)

    @classmethod
    def get(cls, template_id):
        """Retrieve a template. GET /templates/:id"""
        return http_client.request("GET", f"/templates/{_e(template_id)}")

    @classmethod
    def list(cls, params=None):
        """List templates. GET /templates"""
        return http_client.request("GET", f"/templates{paginate(params)}")

    @classmethod
    def update(cls, template_id, params):
        """Update a template. PATCH /templates/:id"""
        return http_client.request("PATCH", f"/templates/{_e(template_id)}", params)

    @classmethod
    def duplicate(cls, template_id, params=None):
        """Duplicate a template. POST /templates/:id/duplicate"""
        return http_client.request(
            "POST",
            f"/templates/{_e(template_id)}/duplicate",
            params if params is not None else {},
        )

    @classmethod
    def publish(cls, template_id):
        """Publish a template (make its latest draft live).
        POST /templates/:id/publish"""
        return http_client.request("POST", f"/templates/{_e(template_id)}/publish")

    @classmethod
    def remove(cls, template_id):
        """Delete a template. DELETE /templates/:id"""
        return http_client.request("DELETE", f"/templates/{_e(template_id)}")
